File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
        int minTillNow = Integer.MAX_VALUE;
        
        int maxProfitTillNow = 0;
        
        for(int stock: prices){
            minTillNow = Math.min(minTillNow, stock);
            maxProfitTillNow = Math.max(maxProfitTillNow, stock-minTillNow);
        }
        
        return maxProfitTillNow;
    }
}

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let l = 0;
    let r = 1;
    let curMaxProfit = 0;
    while (r < prices.length) {
        let curProfit = prices[r] - prices[l];
        if (curProfit < 0) {
            l = r;
        } else {
            curMaxProfit = Math.max(curProfit, curMaxProfit);
        }
        r++;
    }
    return curMaxProfit;
};